inginious/frontend/plugins/upcoming_tasks/__init__.py

```diff
@@ -0,0 +1,141 @@
+
+        # Get last submissions for left panel
+        last_submissions = self.submission_manager.get_user_last_submissions(5, {"courseid": {"$in": list(open_courses.keys())}})
+        except_free_last_submissions = []
+        for submission in last_submissions:
+            try:
+                submission["task"] = open_courses[submission['courseid']].get_task(submission['taskid'])
+                except_free_last_submissions.append(submission)
+            except:
+                pass
+
+        # Get the courses tasks, remove finished ones and courses that have no available unfinished tasks with upcoming deadline in range
+        tasks_data = {}
+        succeeded_courses = []
+        for courseid, course in open_courses.items():
+            tasks = course.get_tasks()
+            outdated_tasks = [taskid for taskid, task in tasks.items() if (not task.get_accessible_time().is_open()) or ((task.get_accessible_time().get_soft_end_date()) > (datetime.now()+timedelta(days=time_planner)))]
+            new_user_task_list = course.get_task_dispenser().get_user_task_list([username])[username]
+            new_user_task_list = [task_id for task_id in new_user_task_list if task_id not in outdated_tasks]
+            tasks_data[courseid] = {taskid: {"succeeded": False, "grade": 0.0} for taskid in new_user_task_list}
+            user_tasks = self.database.user_tasks.find({"username": username, "courseid": course.get_id(), "taskid": {"$in": new_user_task_list}})
+            for user_task in user_tasks:
+                if not user_task["succeeded"]:
+                    tasks_data[courseid][user_task["taskid"]]["succeeded"] = user_task["succeeded"]
+                    tasks_data[courseid][user_task["taskid"]]["grade"] = user_task["grade"]
+                else:
+                    del tasks_data[courseid][user_task["taskid"]]
+            # Remove courses with no unfinished available tasks with deadline and lti courses
+            if (len(tasks_data[courseid]) == 0) or course.is_lti():
+                succeeded_courses.append(courseid)
+        # Remove succeeded courses (including lti courses)
+        for succeeded_course in succeeded_courses:
+            del open_courses[succeeded_course]
+
+        my_render = RenderOrdered()
+        # Sort the courses based on the most urgent task for each course
+        open_courses = OrderedDict( sorted(iter(open_courses.items()), key=lambda x: (sort_by_deadline(x[1], tasks_data[x[0]].keys())[0]).get_accessible_time().get_soft_end_date() ))
+
+        return self.template_helper.render("coming_tasks.html",
+                                           template_folder=PATH_TO_PLUGIN + "/templates/",
+                                           open_courses=open_courses,
+                                           tasks_data=tasks_data,
+                                           my_render=my_render,
```

Comment:
1. Here you can pass the `sort_by_deadline` function in argument instead of `my_render`